File: moviejava/app/src/main/java/com/example/moviejava/ui/home/HomeViewModel.java
package com.example.moviejava.ui.home;


import android.os.Build;

import androidx.annotation.RequiresApi;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.moviejava.data.remote.movie.MovieRepository;
import com.example.moviejava.model.Results;
import com.example.moviejava.model.State;
import com.example.moviejava.ui.base.BaseViewModel;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;

@HiltViewModel
public class HomeViewModel extends BaseViewModel {
    private final MovieRepository repository;

    @RequiresApi(api = Build.VERSION_CODES.N)
    @Inject
    public HomeViewModel(MovieRepository repository) {
        this.repository = repository;
        getUpComing(1);
        getPopular(1);
        getNowPlaying(1);
        getTopRate(1);
    }

    private final List<Results> listUpcoming = new ArrayList<>();
    private final MutableLiveData<State<List<Results>>> _upcoming = new MutableLiveData<>();
    public LiveData<State<List<Results>>> upcoming = _upcoming;

    private final List<Results> listTopRate = new ArrayList<>();
    private final MutableLiveData<State<List<Results>>> _topRate = new MutableLiveData<>();
    public LiveData<State<List<Results>>> topRate = _topRate;

    private final List<Results> listPopular = new ArrayList<>();
    private final MutableLiveData<State<List<Results>>> _popular = new MutableLiveData<>();
    public LiveData<State<List<Results>>> popular = _popular;

    private final List<Results> listNowPlaying = new ArrayList<>();
    private final MutableLiveData<State<List<Results>>> _nowPlaying = new MutableLiveData<>();
    public LiveData<State<List<Results>>> nowPlaying = _nowPlaying;
    private List<Results> listSearch = new ArrayList<>();
    private final MutableLiveData<State<List<Results>>> _search = new MutableLiveData<>();
    public LiveData<State<List<Results>>> search = _search;


    @RequiresApi(api = Build.VERSION_CODES.N)
    public void getUpComing(int page) {
        launchRxSingle(repository.getUpcoming(page), result -> {
            listUpcoming.addAll(result.results);
            _upcoming.postValue(State.success(listUpcoming));
        }, e->android.util.Log.e("ERRR",""+e));
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public void getTopRate(int page) {
        launchRxSingle(repository.getUpcoming(page), result -> {
            listTopRate.addAll(result.results);
            _topRate.postValue(State.success(listTopRate));
        }, e->android.util.Log.e("ERRR",""+e));


    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public void getPopular(int page) {
        launchRxSingle(repository.getUpcoming(page), result -> {
            android.util.Log.e("haiii",""+result);
            listPopular.addAll(result.results);
            _popular.postValue(State.success(listPopular));
        }, e->android.util.Log.e("ERRR",""+e));


    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public void getNowPlaying(int page) {
        launchRxSingle(repository.getUpcoming(page), result -> {
            listNowPlaying.addAll(result.results);
            _nowPlaying.postValue(State.success(listNowPlaying));
        }, e->android.util.Log.e("ERRR",""+e));

    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public void searchMovie(String query, int page, boolean more) {
        launchRxSingle(repository.getUpcoming(page), result -> {
            if (more) {
                listSearch.addAll(result.results);
            }
            listSearch = result.results;

            _search.postValue(State.success(listSearch));
        }, e->android.util.Log.e("ERRR",""+e));
    }
}

File: movie/app/src/main/java/com/example/movie/ui/home/HomeViewModel.kt
package com.example.movie.ui.home

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.example.movie.data.remote.movie.MovieRepository
import com.example.movie.model.Results
import com.example.movie.model.State
import com.example.movie.ui.base.BaseViewModel
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject

@HiltViewModel
class HomeViewModel @Inject constructor(private val repository: MovieRepository) : BaseViewModel() {

    private var listUpcoming = mutableListOf<Results>()
    private val _upcoming = MutableLiveData<State<List<Results>>>()
    val upcoming: LiveData<State<List<Results>>>
        get() = _upcoming

    private var listTopRate = mutableListOf<Results>()
    private val _topRate = MutableLiveData<State<List<Results>>>()
    val topRate: LiveData<State<List<Results>>>
        get() = _topRate

    private var listPopular = mutableListOf<Results>()
    private val _popular = MutableLiveData<State<List<Results>>>()
    val popular: LiveData<State<List<Results>>>
        get() = _popular

    private var listNowPlaying = mutableListOf<Results>()
    private val _nowPlaying = MutableLiveData<State<List<Results>>>()
    val nowPlaying: LiveData<State<List<Results>>>
        get() = _nowPlaying
    private var listSearch = mutableListOf<Results>()
    private val _search = MutableLiveData<State<List<Results>>>()
    val search: LiveData<State<List<Results>>>
        get() = _search

    init {
        getUpComing(page = 1)
        getPopular(page = 1)
        getNowPlaying(page = 1)
        getTopRate(page = 1)

    }

    fun getUpComing(page: Int) {
        _upcoming.postValue(State.loading())
        launchViewModel(repository.getUpcoming(page = page), {
            listUpcoming.addAll(it.results)
            _upcoming.postValue(State.success(listUpcoming))
        }, {
            Log.e("err", it.message.toString())
        })
    }

    fun getTopRate(page: Int) {
        _topRate.postValue(State.loading())
        launchViewModel(repository.getTopRate(page = page), {
            listTopRate.addAll(it.results)
            _topRate.postValue(State.success(listTopRate))
        }, {
            Log.e("err", it.message.toString())
        })
    }

    fun getPopular(page: Int) {
        _popular.postValue(State.loading())
        launchViewModel(repository.getPopular(page = page), {
            listPopular.addAll(it.results)
            _popular.postValue(State.success(listPopular))
        }, {
            Log.e("err", it.message.toString())
        })


    }

    fun getNowPlaying(page: Int) {
        _nowPlaying.postValue(State.loading())
        launchViewModel(repository.getNowPlaying(page = page), {
            listNowPlaying.addAll(it.results)
            _nowPlaying.postValue(State.success(listNowPlaying))
        }, {
            Log.e("err", it.message.toString())
        })

    }

    fun searchMovie(query: String, page: Int, more: Boolean) {

        launchViewModel(repository.searchMovie(query = query, page = page), {
            if (more) listSearch.addAll(it.results)
            else listSearch = it.results
            _search.postValue(State.success(listSearch))
        }, {
            Log.e("err", it.message.toString())
        })


    }
}